### PR TITLE
Integrate mycelium in zos

### PR DESCRIPTION
**Pull request by Eslam-Nawara**
### Description

Integrate mycelium in zos and allow flists to be served over mycelium

### Changes

Add mycelium to networkd

### Related Issues

- #2338 

### Checklist

- [ ] Tests included
- [x] Build pass
- [x] Documentation
- [x] Code format and docstring

### CODE REVIEW

**Comment by xmonader** [comment]
don't forget to add an ADR please

File: pkg/flist/flist.go [hunk 1]
```diff
@@ -190,13 +193,13 @@ func New(root string, storage *stubs.StorageModuleStub) pkg.Flister {
...
 // TODO: how to know that this ro mount is no longer used, hence can be unmounted and cleaned up?
-func (f *flistModule) mountRO(url, storage string) (string, error) {
+func (f *flistModule) mountRO(url, storage, namespace string) (string, error) {
 	// this should return always the flist mountpoint. which is used
 	// as a base for all RW mounts.
 	sublog := log.With().Str("url", url).Str("storage", storage).Logger()
 	sublog.Info().Msg("request to mount flist")
 
-	hash, flistPath, err := f.downloadFlist(url)
+	hash, flistPath, err := f.downloadFlist(url, namespace)
```
Comment (muhamadazmy) [inline pkg/flist/flist.go]: You probably need to use this same namespace at line `262` below

File: cmds/modules/networkd/main.go [hunk 2]
```diff
@@ -126,6 +128,17 @@ func action(cli *cli.Context) error {
 		return errors.Wrap(err, "failed to start yggdrasil")
 	}
 
+	log.Debug().Msg("starting mycelium")
```
Comment (xmonader) [inline cmds/modules/networkd/main.go:126]: I think this and ygg code can be extracted into their own functions and same for the DMZ

File: cmds/modules/networkd/main.go [hunk 3]
```diff
@@ -148,9 +161,34 @@ func action(cli *cli.Context) error {
 		if err := dmzYgg.SetYggIP(ip, gw.IP); err != nil {
 			return errors.Wrap(err, "failed to set yggdrasil ip for dmz")
 		}
+
+		// set up mycelium in ndmz
+		dmzMy, err := mycelium.NewMyNamespace(dmz.Namespace())
```
Comment (xmonader) [inline cmds/modules/networkd/main.go]: can be extracted into another function

File: docs/architecture/decisions/0010-mycelium.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+Accepted
+
+## Context
+
+Support mycelium network for zdbs and zos host to serve flists
+
+## Decision
+
+Integrate mycelium in zos and allow zos host to have mycelium IPs, serve flists over mycelium, and support mycelium on zdbs
```
Comment (xmonader) [inline docs/architecture/decisions/0010-mycelium.md]: is it serve or mount flists over mycelium?

File: pkg/network/types/constant.go [hunk 5]
```diff
@@ -11,8 +11,10 @@ const (
...
 	// by the bootstrap of networkd
 	DefaultBridge = "zos"
-	//PublicBridge name
+	// PublicBridge name
 	PublicBridge = "br-pub"
-	//YggBridge ygg bridge
+	// YggBridge ygg bridge
 	YggBridge = "br-ygg"
+	// MyBridge mycelium bridge
+	MyBridge = "br-my"
```
Comment (xmonader) [inline pkg/network/types/constant.go]: MyCeliumBridge

File: pkg/network/networker.go [hunk 6]
```diff
@@ -192,6 +196,48 @@ func (n *networker) detachYgg(id string, netNs ns.NetNS) error {
...
+	// new hardware address for mycelium interface
+	hw := ifaceutil.HardwareAddrFromInputBytes([]byte("my:" + id))
+
+	myc, err := n.myc.InspectMycelium()
+	if err != nil {
+		return net.IPNet{}, err
+	}
+
+	ip, err := myc.SubnetFor(hw)
+	log.Info().Msgf("myc subnet is %s", ip)
```
Comment (xmonader) [inline pkg/network/networker.go]: myc -> mycelium

File: pkg/network/networker.go [hunk 7]
```diff
@@ -114,13 +117,14 @@ func NewNetworker(identity *stubs.IdentityManagerStub, ndmz ndmz.DMZ, ygg *yggdr
 		portSet:        set.NewInt(),
 
 		ygg:  ygg,
+		myc:  myc,
```
Comment (xmonader) [inline pkg/network/networker.go]: mycelium:

File: pkg/network/mycelium/mycelium.go [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+	f, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	key := x25519.EdPrivateKeyToX25519([]byte(privateKey))
+
+	return json.NewEncoder(f).Encode(key)
```
Comment (muhamadazmy) [inline pkg/network/mycelium/mycelium.go]: Why are u using `json` here. the priv key of mycelium is not json encoded. It's a binary file. i don't believe this file is loaded by mycelium and is either ignored is overridden.

If this ever worked, then i assume it's because mycelium just read key length bytes from the file and used that as seed for the key!

This is definitely an issue, because it means the key u generated here is not secure. because it's not the full "Random" key but only part of the key json encoded bytes.

Do you get what I mean ?

File: pkg/network/mycelium/utils.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+import (
+	"context"
+	"crypto/ed25519"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/zos/pkg/zinit"
+)
+
+func EnsureMycelium(ctx context.Context, privateKey ed25519.PrivateKey, ns MyceliumNamespace) (*MyServer, error) {
```
Comment (xmonader) [inline pkg/network/mycelium/utils.go]: move this into mycelium.go and remove utils file

File: pkg/network/mycelium/config.go [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+	"github.com/decred/base58"
+	"github.com/pkg/errors"
+)
+
+type NodeConfig struct {
+	KeyFile    string
+	TunName    string
+	Peers      []string
+	PrivateKey ed25519.PrivateKey
+	PublicKey  string
```
Comment (muhamadazmy) [inline pkg/network/mycelium/config.go]: Why is this needed?

File: pkg/network/mycelium/config.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+	"github.com/decred/base58"
+	"github.com/pkg/errors"
+)
+
+type NodeConfig struct {
+	KeyFile    string
+	TunName    string
+	Peers      []string
+	PrivateKey ed25519.PrivateKey
```
Comment (muhamadazmy) [inline pkg/network/mycelium/config.go]: Why do you keep the private key as ed25519 while we need x25519. I think it's better to convert the key before storing it on the NodeConfig from the start instead of changing it on start

File: pkg/network/mycelium/config.go [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+	PublicKey  string
+	NodeInfo   map[string]interface{}
```
Comment (muhamadazmy) [inline pkg/network/mycelium/config.go]: What is this and what is it used for?

We don't need to copy everything over from ygg. Ygg has a completely different configuration and even requires a config file with way more config than mycelium.

Please drop all attributes that are not needed.

File: pkg/network/mycelium/mycelium.go [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+	"github.com/threefoldtech/zos/pkg/network/namespace"
+	"github.com/threefoldtech/zos/pkg/zinit"
+)
+
+const (
+	tunName      = "my0"
+	myBin        = "mycelium"
+	zinitService = "mycelium"
+	confPath     = "/tmp/mycelium_priv_key.bin"
+	MyListenTCP  = 9651
```
Comment (muhamadazmy) [inline pkg/network/mycelium/mycelium.go:28]: you will have to add this to the list of reserved ports of the system

File: pkg/network/mycelium/mycelium.go [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+}
+
+// NodeID returns mycelium node ID of s
+func (s *MyServer) NodeID() (ed25519.PublicKey, error) {
+	if s.cfg.PublicKey == "" {
+		panic("EncryptionPublicKey empty")
+	}
+
+	return hex.DecodeString(s.cfg.PublicKey)
+}
```
Comment (muhamadazmy) [inline pkg/network/mycelium/mycelium.go]: Not needed

File: pkg/network/mycelium/mycelium.go [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+	if err != nil {
+		return inspection, err
+	}
+
+	bin, err := exec.LookPath(myBin)
+	if err != nil {
+		return inspection, err
+	}
+
+	cmd := fmt.Sprintf(`exec ip netns exec %s %s inspect --json --key-file %s `, s.ns, bin, confPath)
```
Comment (muhamadazmy) [inline pkg/network/mycelium/mycelium.go]: Any idea why your command is a `sh` then u do exec and not the command directly!
I mean why not do 
```
os.Command("ip", "netns", "exec", s.ns,  bin, "inspect", "--json", "--key-file", confPath)
```
?!

File: pkg/network/mycelium/mycelium.go [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+	cfg *NodeConfig
+	ns  string
+}
+
+type MyceliumInspection struct {
+	PublicKey string `json:"publicKey"`
+	Address   net.IP `json:"address"`
+}
+
+type Subnet [8]byte
```
Comment (muhamadazmy) [inline pkg/network/mycelium/mycelium.go]: this is not used

File: pkg/network/mycelium/mycelium.go [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+}
+
+// Tun return the name of the TUN interface created by mycelium
+func (s *MyServer) Tun() string {
+	return s.cfg.TunName
+}
+
+// SubnetFor return an IP address out of the node allocated subnet by hasing b and using it
+// to generate the last 64 bits of the IPV6 address
+func (m *MyceliumInspection) SubnetFor(b []byte) (net.IPNet, error) {
```
Comment (muhamadazmy) [inline pkg/network/mycelium/mycelium.go]: I really think this should be called `IPFor` not `SubnetFor` even if it's called like that in ygg it's still wrong.

**Review by muhamadazmy — CHANGES_REQUESTED** [review]

File: pkg/network/mycelium/mycelium.go (hunk 13 above)
Comment (Eslam-Nawara) [reply pkg/network/mycelium/mycelium.go:28]: it's added to reserved ports [here](https://github.com/threefoldtech/zos/pull/2342/files/29253a863afba63ab9b040e4e464e26a970bf775#diff-4e15b6f2760eb52c8292eadf92ce4a6ba3afcf08d34b528d56081b9bc1623554R127)

**Review by muhamadazmy — APPROVED** [review]
